### PR TITLE
Lower layers loop

### PR DESCRIPTION
**Pull request by doru1004**
Add more AST node types.

### CODE REVIEW

File: coffee/base.py [hunk 1]
```diff
@@ -663,6 +731,24 @@ def gencode(self, not_scope=False):
...
+        # If the then branch is a plain list, cast it to a Block.
+        if not isinstance(then_body, Node):
+            then_body = Block(then_body, open_scope=True)
+        super(Ifthen, self).__init__([then_body])
+        self.cond = cond
+
+    def gencode(self, scope=True):
+        return "if (" + str(self.cond) + ")\n" + indent(self.children[0].gencode())
+
+
```
Comment (wence-) [inline coffee/base.py]: OK, I see you renamed to Ifthen, from If.  But I'm confused why you need this?

File: coffee/base.py [hunk 2]
```diff
@@ -216,6 +216,30 @@ def gencode(self, not_scope=True):
...
+class Addr(UnaryExpr):
+
+    """Get the address."""
+
+    def gencode(self):
+        return "&%s" % wrap(self.children[0].gencode())
+
+
```
Comment (dorugeber) [inline coffee/base.py]: COFFEE is just turning into verbose-C?

File: coffee/base.py [hunk 3]
```diff
@@ -216,6 +216,30 @@ def gencode(self, not_scope=True):
...
+    def gencode(self):
+        return "*%s" % wrap(self.children[0].gencode())
+
+
+class Cast(UnaryExpr):
+
+    """Dereference an address."""
+
+    def gencode(self):
+        return "(%s) %s" % wrap(self.children[0].gencode())
```
Comment (wence-) [inline coffee/base.py]: Does wrap return two strings?  I can't understand how this works?

File: coffee/base.py (hunk 1 above)
Comment (doru1004) [inline coffee/base.py]: Initially there was no IF and I added an If with only the then branch present. In the meantime another If was added with both then and else branches. Do you want me to code this IfThen into the If node?

File: coffee/base.py (hunk 2 above)
Comment (doru1004) [inline coffee/base.py]: Well we have to support all the unary operators supported by C.

File: coffee/utils.py [hunk 4]
```diff
@@ -643,3 +644,72 @@ def set_itspace(loop_a, loop_b):
...
+        if var_name == old_name:
+            kernel_ast = scopy(kernel_ast)
+            kernel_ast.set_itvar(new_name)
+    children = []
+    for child in kernel_ast.children:
+        children.append(rename_var(child, old_name, new_name))
+    kernel_ast.children = children
+    return kernel_ast
+
+def contains_flat_block(tree):
```
Comment (FabioLuporini) [inline coffee/utils.py]: You don't need this, just use visit(node, search=FlatBlock)

File: coffee/utils.py [hunk 5]
```diff
@@ -643,3 +644,72 @@ def set_itspace(loop_a, loop_b):
...
+    # Check if the AST contains any FlatBlock nodes
+    if isinstance(tree, FlatBlock):
+        return True
+    elif not isinstance(tree, str):
+        for child in tree.children:
+            if contains_flat_block(child):
+                return True
+    return False
+
+def change_kernel_loop_index_ids(tree):
```
Comment (FabioLuporini) [inline coffee/utils.py]: I think there's already another utility function for doing this

File: coffee/utils.py [hunk 6]
```diff
@@ -643,3 +644,72 @@ def set_itspace(loop_a, loop_b):
 def loops_as_dict(loops):
     loops_itvars = [l.itvar for l in loops]
     return OrderedDict(zip(loops_itvars, loops))
+
+def rename_var(kernel_ast, old_name, new_name):
```
Comment (FabioLuporini) [inline coffee/utils.py]: Can't you just use the utility function ast_replace?

File: coffee/utils.py [hunk 7]
```diff
@@ -643,3 +644,72 @@ def set_itspace(loop_a, loop_b):
...
+    This method renames the variables represented by a Symbol object.
+    The Symbol's string is replaces the ``old_name`` with the
+    new string ``new_name``.
+    """
+    # Warning: Inconsistencies in the representation could make this
+    # function fail if variable names are polluted with
+    # *, & and [] symbols.
+    if isinstance(kernel_ast, Symbol):
+        var_name = str(kernel_ast.symbol)
+        if "*" in var_name or "[" in var_name or "]" in var_name or "&" in var_name:
```
Comment (FabioLuporini) [inline coffee/utils.py]: We have to find a better way of doing this

File: coffee/utils.py [hunk 8]
```diff
@@ -643,3 +644,72 @@ def set_itspace(loop_a, loop_b):
...
+        return kernel_ast
+    if not kernel_ast.has_children:
+        return kernel_ast
+    if isinstance(kernel_ast, For):
+        var_name = str(kernel_ast.itvar)
+        if "*" in var_name or "[" in var_name or "]" in var_name or "&" in var_name:
+            raise RuntimeError("The for loop index name name has been polluted with * & [ or ] characters.")
+        if var_name == old_name:
+            kernel_ast = scopy(kernel_ast)
+            kernel_ast.set_itvar(new_name)
```
Comment (FabioLuporini) [inline coffee/utils.py]: Also, looks like copy-paste

File: coffee/base.py [hunk 9]
```diff
@@ -683,22 +749,30 @@ def gencode(self):
...
+    """If-then-else construct.
 
     :param if_expr: The expression driving the jump
-    :param branches: A 2-tuple of AST nodes, respectively the 'if' and the 'else'
+    :param branches: A 2-tuple of AST nodes, respectively the 'then' and the 'else'
                      branches
     """
 
     def __init__(self, if_expr, branches):
+        # TODO: Re-implement this using two arguments a then branch and an optional else branch
```
Comment (FabioLuporini) [inline coffee/base.py]: I was avoiding TODOs in master... What are you trying to do here?

File: coffee/base.py [hunk 10]
```diff
@@ -683,22 +749,30 @@ def gencode(self):
...
+        self.then_body = branches[0]
+        if not isinstance(branches[0], Node):
```
Comment (FabioLuporini) [inline coffee/base.py]: Why shouldn't they be Nodes?

File: coffee/base.py [hunk 11]
```diff
@@ -683,22 +749,30 @@ def gencode(self):
...
+            self.then_body = Block(branches[0], open_scope=True)
+        if len(branches) == 2:
+            self.else_body = branches[1]
+            if not isinstance(branches[1], Node):
```
Comment (FabioLuporini) [inline coffee/base.py]: Why shouldn't they be Nodes?

File: coffee/base.py [hunk 12]
```diff
@@ -632,6 +672,32 @@ def itvar(self):
...
+        if isinstance(self.cond.children[0], Symbol):
+            self.cond = scopy(self.cond)
+            self.cond.children[0].symbol = new_var
+        elif isinstance(self.cond.children[0], str):
+            self.cond.children[0] = new_var
+        else:
+            raise NotImplemented("Don't know how to change index name in this condition node.")
+        # Change the name of the index in the increment
+        if isinstance(self.incr.children[0], Symbol):
+            self.incr = scopy(self.incr)
```
Comment (FabioLuporini) [inline coffee/base.py]: shallow copies are dangerous. Why are you using them?

File: coffee/base.py [hunk 13]
```diff
@@ -632,6 +672,32 @@ def itvar(self):
...
+    def set_itvar(self, new_var):
+        # Change the name of the index init
+        self.init = scopy(self.init)
+        if isinstance(self.init, Decl):
+            self.init.sym.symbol = new_var
+        elif isinstance(self.init, Assign):
+            self.init = Decl("int", Symbol(new_var), self.init.children[1])
+        else:
+            raise NotImplemented("Don't know how to change index name in this init node.")
+        # Change the name of the index in the comparison
```
Comment (FabioLuporini) [inline coffee/base.py]: Why are you checking between Symbol and str? str are *not* expected at this point. If there are, that should be fixed and turned into Symbols

**Comment by FabioLuporini** [comment]
Apart from the various comments I made, this is not rebased on master

**Comment by FabioLuporini** [comment]
This can now be rebased on master

**Comment by FabioLuporini** [comment]
@doru1004  any plans for this?

**Comment by miklos1** [comment]
I'm closing this now, because it has been abandoned for more than a year. If it is still relevant, please rebase and reopen.

**Pull request closed without merging**